src/test/java/org/apache/commons/codec/binary/Base16Test.java

```diff
@@ -0,0 +1,625 @@
+    }
+
+    @Test
+    public void testByteToStringVariations() throws DecoderException {
+        final Base16 Base16 = new Base16();
+        final byte[] b1 = StringUtils.getBytesUtf8("Hello World");
+        final byte[] b2 = new byte[0];
+        final byte[] b3 = null;
+
+        assertEquals("byteToString Hello World", "48656C6C6F20576F726C64", Base16.encodeToString(b1));
+        assertEquals("byteToString static Hello World", "48656C6C6F20576F726C64", StringUtils.newStringUtf8(new Base16().encode(b1)));
+        assertEquals("byteToString \"\"", "", Base16.encodeToString(b2));
+        assertEquals("byteToString static \"\"", "", StringUtils.newStringUtf8(new Base16().encode(b2)));
+        assertEquals("byteToString null", null, Base16.encodeToString(b3));
+        assertEquals("byteToString static null", null, StringUtils.newStringUtf8(new Base16().encode(b3)));
+    }
+
+    @Test
+    public void testStringToByteVariations() throws DecoderException {
+        final Base16 Base16 = new Base16();
+        final String s1 = "48656C6C6F20576F726C64";
+        final String s2 = "";
+        final String s3 = null;
+
+        assertEquals("StringToByte Hello World", "Hello World", StringUtils.newStringUtf8(Base16.decode(s1)));
+        assertEquals("StringToByte Hello World", "Hello World",
+                StringUtils.newStringUtf8((byte[]) new Base16().decode((Object) s1)));
+        assertEquals("StringToByte static Hello World", "Hello World",
+                StringUtils.newStringUtf8(new Base16().decode(s1)));
+        assertEquals("StringToByte \"\"", "", StringUtils.newStringUtf8(new Base16().decode(s2)));
+        assertEquals("StringToByte static \"\"", "", StringUtils.newStringUtf8(new Base16().decode(s2)));
+        assertEquals("StringToByte null", null, StringUtils.newStringUtf8(new Base16().decode(s3)));
+        assertEquals("StringToByte static null", null, StringUtils.newStringUtf8(new Base16().decode(s3)));
+    }
+
+    private String toString(final byte[] data) {
+        final StringBuilder buf = new StringBuilder();
+        for (int i = 0; i < data.length; i++) {
+            buf.append(data[i]);
+            if (i != data.length - 1) {
+                buf.append(",");
+            }
+        }
+        return buf.toString();
+    }
+
+    /**
+     * Test for CODEC-265: Encode a ~1GiB file.
+     *
+     * @see <a href="https://issues.apache.org/jira/projects/CODEC/issues/CODEC-265">CODEC-265</a>
+     */
+    public void testCodec265_over() {
+        // almost 1GiB file to encode: 2^29 bytes
+        final int size1GiB = 1 << 29;
+
+        // Expecting a size of 2 output bytes per 1 input byte
+        final int blocks = size1GiB;
+        final int expectedLength = 2 * blocks;
+
+        // This test is memory hungry. Check we can run it.
+        final long presumableFreeMemory = BaseNCodecTest.getPresumableFreeMemory();
+
+        // Estimate the maximum memory required:
```

Comment:
This has been copied from the Base64Test, changed (to encode a 500MiB file) and then disabled (no `@Test` annotation).

Codec 265 concerns encoding a 1GiB file. Using base 64 you get 4 characters per 3 bytes so the output is ~1.33GiB. With Base16 you get 2 characters per byte and it is not possible to create a 2 GiB char array.

Note: It will never run on Travis CI due to memory restrictions. When I run this locally it passes with a 500MiB input (1 << 29) but fails with 1GiB (1 << 30). However this is not due to an out-of-memory error (as expected) but an index out of bounds exception. This is due to the allocation of the buffer using:
```java
context.buffer = new byte[Math.max(size, getDefaultBufferSize())];
```
As noted in my review comment for `ensureBufferSize` this has been changed to allow a single allocation of the entire array. Unfortunately when size is negative this creates a buffer of the default size.

Given that `BaseNCodec.ensureBufferSize` is intended to be called with a positive-only size one suggestion is `Base16.encode` could be updated to change this line:
```java
final byte[] buffer = ensureBufferSize(length * BYTES_PER_ENCODED_BLOCK, context)
```
to
```java
final int size = length * BYTES_PER_ENCODED_BLOCK;
if (size < 0) {
    throw new IllegalArgumentException("Input byte[] length exceeds maximum size for encoded data: " + length);
}

final byte[] buffer = ensureBufferSize(size, context)
```

A different approach with a less meaningful exception (a negative array size exception) would be to change ensureBufferSize:
```java
context.buffer = new byte[
    compareUnsigned(size, getDefaultBufferSize()) > 0
       ? size
       : getDefaultBufferSize()];
```

Either way a test should be added to hit this exception.

Reply:
I see that the change to `Base16.encode` to handle overflow has been done but this test still exists. It should be replaced by an attempt to encode a `byte[]` that is length 2^30. This will require 1Gb of memory and should be OK on Travis CI. Something like:
```java
@Test(expected = IllegalArgumentException.class)
public void testMaximumEncodedSizeExceeded() {
  final int size1GiB = 1 << 30;
  final byte[] bytes = new byte[size1GiB];
  new Base16().encode(bytes);
}
```

Reply:
@aherbert I already added such a test as you requested. See the test function directly below your comment named `checkEncodeLengthBounds`

Reply:
OK. It uses a different method call where the length is a separate argument but given the simplicity of the encode method there is no point in testing `encode(new byte[1 << 30])` as well.  So just remove `testCodec265_over()` as the test makes no sense for Base16.